### PR TITLE
Enable SG/SGAccel windows install/uninstall

### LINKED ISSUE #1277: Automated SG/SGAccel windows install/uninstall
The current provisioning works fine for centos. We need the install/uninstall to also work for Windows so that the SG functional tests can also be run against windows.

### PR DESCRIPTION
**Pull request by raghusarangapani**
#### Fixes #1277.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:
- Enables SG/SGAccel support for Windows

### CODE REVIEW

**Comment by sethrosetter** [comment]
Why do we have to do individual `becomes`? It was my understanding that this should be inherited to all tasks if this is defined at the playbook scope?

**Comment by raghusarangapani** [comment]
At the playbook scope, now we call centos specific task which has the become or a windows specific task which does not have the become. `become` is invalid for windows. If I put it at the playbook level, the windows run will error out.

File: libraries/provision/ansible/playbooks/install-sync-gateway-package.yml [hunk 1]
```diff
@@ -57,10 +85,14 @@
...
 
   tasks:
-  - debug: msg="Downloading sync_gateway v. {{ couchbase_sync_gateway_package_url }}"
-  - name: SYNC GATEWAY |  Download sync_gateway rpm
+  - name: SYNC GATEWAY |  Download sync_gateway rpm {{ couchbase_sync_gateway_package_url }}
     get_url: url={{ couchbase_sync_gateway_package_url }} dest=/tmp/{{ couchbase_sync_gateway_package }}
+    when: ansible_distribution == "CentOS"
 
+  - name: SYNC GATEWAY |  Download sync_gateway exe {{ couchbase_sync_gateway_package_url | regex_replace('rpm', 'exe') }}
+    win_get_url: url={{ couchbase_sync_gateway_package_url | regex_replace('rpm', 'exe') }} dest="C:\Users\Administrator\AppData\Local\Temp\{{ couchbase_sync_gateway_package | regex_replace('(.*)rpm$', '\\1exe') }}"
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/install-sync-gateway-package.yml]: I think we should redesign how we pass the package name in. This does not seem scalable to other package types

File: libraries/provision/ansible/playbooks/install-sync-gateway-package.yml [hunk 2]
```diff
@@ -1,16 +1,44 @@
...
+    couchbase_sync_gateway_package_base_url:
+    couchbase_sync_gateway_package:
+    couchbase_sync_gateway_package_url: "{{ couchbase_sync_gateway_package_base_url }}/{{ couchbase_sync_gateway_package }}"
 
   tasks:
+  - include: tasks/stop-sync-gateway.yml
+    when: ansible_distribution == "CentOS"
+
+  - include: tasks/stop-sync-gateway-windows.yml
+    when: ansible_os_family == "Windows"
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/install-sync-gateway-package.yml]: Is it possible to refactor this so that all the windows / centos ops are in a separate playbooks?

File: libraries/provision/ansible/playbooks/install-sync-gateway-package.yml [hunk 3]
```diff
@@ -72,14 +104,37 @@
...
 
   tasks:
-  - debug: msg="Downloading sg_accel v. {{ couchbase_sg_accel_package_url }}"
-  - name: SYNC GATEWAY | Download sg_accel rpm
+  - name: SG Accel | Download sg_accel rpm {{ couchbase_sg_accel_package_url }}
     get_url: url={{ couchbase_sg_accel_package_url }} dest=/tmp/{{ couchbase_sg_accel_package }}
+    when: ansible_distribution == "CentOS"
+
+  - name: SG Accel | Download sg_accel exe {{ couchbase_sg_accel_package_url | regex_replace('rpm', 'exe') }}
+    win_get_url: url={{ couchbase_sg_accel_package_url | regex_replace('rpm', 'exe') }} dest="C:\Users\Administrator\AppData\Local\Temp\{{ couchbase_sg_accel_package | regex_replace('(.*)rpm$', '\\1exe') }}"
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/install-sync-gateway-package.yml]: Same comment as above

File: libraries/provision/ansible/playbooks/install-sync-gateway-package.yml [hunk 4]
```diff
@@ -72,14 +104,37 @@
...
+  - include: tasks/install-sg-accel.yml
+    when: ansible_distribution == "CentOS"
+
+  - include: tasks/install-sg-accel-windows.yml
+    when: ansible_os_family == "Windows"
 
 # Deploy non writer sync_gateway configs
 - hosts: sync_gateways
   any_errors_fatal: true
-  become: yes
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/install-sync-gateway-package.yml]: Why is this removed now?

File: libraries/provision/ansible/playbooks/remove-previous-installs.yml [hunk 5]
```diff
@@ -1,27 +1,37 @@
 ---
 - hosts: pool
   any_errors_fatal: true
-  become: yes
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/remove-previous-installs.yml]: why is this removed now?

File: libraries/provision/ansible/playbooks/tasks/install-sg-accel-windows.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Install and stop sg_accel service
+- name: SG ACCEL | Install sg_accel exe {{ couchbase_sg_accel_package | regex_replace('rpm', 'exe') }}
+  win_shell: C:\Users\Administrator\AppData\Local\Temp\{{ couchbase_sg_accel_package | regex_replace('rpm', 'exe') }} /S /v /qn
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/tasks/install-sg-accel-windows.yml]: Same as above

File: libraries/provision/ansible/playbooks/tasks/install-sg-windows.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Install and stop sync_gateway service
+- name: SYNC GATEWAY | Install sync_gateway exe {{ couchbase_sync_gateway_package | regex_replace('rpm', 'exe') }}
+  win_shell: C:\Users\Administrator\AppData\Local\Temp\{{ couchbase_sync_gateway_package | regex_replace('rpm', 'exe') }} /S /v /qn
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/tasks/install-sg-windows.yml]: same as above

File: libraries/provision/ansible/playbooks/tasks/start-sg-accel-windows.yml [hunk 8]
```diff
@@ -3,4 +3,13 @@
 - name: SG ACCEL | Start sg_accel service on Windows
   win_service:
     name: SGAccel
-    state: started
\ No newline at end of file
+    state: started
+
+# There is no win_wait_for
+# This workaround delegates port check to localhost
```
Comment (sethrosetter) [inline libraries/provision/ansible/playbooks/tasks/start-sg-accel-windows.yml]: What do you mean by this?

File: libraries/utilities/generate_clusters_from_pool.py [hunk 9]
```diff
@@ -363,6 +363,22 @@ def write_config(config, pool_file, use_docker):
...
+            f.write("ansible_connection=winrm\n")
+            f.write("ansible_winrm_server_cert_validation=ignore\n")
+
+        if sg_accel_windows:
+            f.write("\n\n[sg_accels:vars]\n")
+            f.write("ansible_user=Administrator\n")
+            f.write("ansible_password=Membase123\n")
+            f.write("ansible_port=5986\n")
+            f.write("ansible_connection=winrm\n")
+            f.write("ansible_winrm_server_cert_validation=ignore\n")
```
Comment (sethrosetter) [inline libraries/utilities/generate_clusters_from_pool.py]: Just so I'm clear, the config generation with append these values to the ansible config and assume that all of the sg / sg_accels are windows?

**Comment by raghusarangapani** [comment]
The functional tests on centos with this branch are here:

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-cc/1272/
- 100% pass
http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/1095/
- 1 failure, unrelated to this change
